src/mlpack/methods/ann/layer/neural_turing_machine_impl.hpp

```diff
@@ -0,0 +1,367 @@
+          readHead) * arma::trans(boost::apply_visitor(deltaVisitor,
+          controller).submat(inSize, 0, inSize + memSize - 1, 0));
+
+      // Backward pass through read head.
+      boost::apply_visitor(BackwardVisitor(std::move(
+          boost::apply_visitor(outputParameterVisitor, readHead)),
+          std::move(dReadHead), std::move(boost::apply_visitor(deltaVisitor,
+          readHead))), readHead);
+
+      prevError = gy + boost::apply_visitor(deltaVisitor, readHead);
+    }
+  }
+
+  // Backward through controller.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, controller)), std::move(prevError),
+      std::move(boost::apply_visitor(deltaVisitor, controller))),
+      controller);
+
+  // Return the delta of the input
+  g = boost::apply_visitor(deltaVisitor, controller).submat(0, 0,
+      inSize - 1, 0);
+
+  bMemoryHistory--;
+  backwardStep++;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NeuralTuringMachine<InputDataType, OutputDataType>::Gradient(
+    arma::Mat<eT>&& input,
+    arma::Mat<eT>&& /* error */,
+    arma::Mat<eT>&& /* gradient */)
+{
+  if (gReads == lReads.end())
+  {
+    gReads = --(--lReads.end());
+
+    gradientStep = 0;
+  }
+  else
+  {
+    // Gradient of read head.
+    boost::apply_visitor(GradientVisitor(std::move(boost::apply_visitor(
+        outputParameterVisitor, controller)),
+        std::move(dReadHead)),
+        readHead);
+
+    if (gradientStep > 1)
+    {
+      // Gradient of linear layer.
+      boost::apply_visitor(GradientVisitor(std::move(boost::apply_visitor(
+          outputParameterVisitor, controller)),
+          std::move(linearError)),
+          inputToLinear);
+
+      // Gradient of write head.
+      boost::apply_visitor(GradientVisitor(std::move(boost::apply_visitor(
+          outputParameterVisitor, controller)),
+          std::move(dWriteHead)),
+          writeHead);
+    }
+  }
+
+  // Gradient of the controller
+  boost::apply_visitor(GradientVisitor(std::move(arma::join_vert(input,
+      *gReads)), std::move(prevError)),
+      controller);
+
+  gReads--;
+  gradientStep++;
+}
+
+template<typename InputDataType, typename OutputDataType>
+void NeuralTuringMachine<InputDataType, OutputDataType>::ResetCell()
+{
+  memoryHistory.clear();
+
+  memoryHistory.push_back(std::move(arma::mat(allOnes.memptr(),
+      allOnes.n_rows, allOnes.n_cols, false, true)));
+
+  bMemoryHistory = memoryHistory.end();
+
+  lReads.clear();
+
+  lReads.push_back(std::move(arma::mat(allZeros.memptr(),
+      allZeros.n_rows, allZeros.n_cols, false, true)));
+
+  gReads = lReads.end();
+
+  backwardStep = 0;
+  gradientStep = 0;
+  forwardStep = 0;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename Archive>
+void NeuralTuringMachine<InputDataType, OutputDataType>::Serialize(
+    Archive& ar, const unsigned int /* version */)
+{
```

Comment:
Hm, I think that we actually do need to do serialization here, can you handle that?  You can take a look at the PR I just opened: #1103.  (Actually it would be good if you looked at that anyway, I made some changes to `GRU`.)

Reply:
Sure. I will add the changes according to your changes to GRU.